livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/assistant_llm.py

```diff
@@ -0,0 +1,246 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import asyncio
+import uuid
+from dataclasses import dataclass, field
+from typing import Any, MutableSet
+
+import httpx
+from livekit.agents import llm
+
+from openai import AsyncAssistantEventHandler, AsyncClient
+from openai.types.beta.threads import Text, TextDelta
+from openai.types.beta.threads.runs import ToolCall, ToolCallDelta
+
+from .models import AssistantTools, ChatModels
+from .utils import build_oai_message
+
+DEFAULT_MODEL = "gpt-4o"
+MESSAGE_ID_KEY = "__openai_message_id__"
+MESSAGES_ADDED_KEY = "__openai_messages_added__"
+
+
+@dataclass
+class LLMOptions:
+    model: str | ChatModels
+
+
+@dataclass
+class AssistantCreateOptions:
+    name: str
+    instructions: str
+    model: ChatModels
+    tools: list[AssistantTools] = field(default_factory=list)
+
+
+@dataclass
+class AssistantLoadOptions:
+    assistant_id: str
+    thread_id: str
+
+
+class AssistantLLM(llm.LLM):
+    def __init__(
+        self,
+        *,
+        assistant_opts: AssistantCreateOptions | AssistantLoadOptions,
+        client: AsyncClient | None = None,
+        api_key: str | None = None,
+        base_url: str | None = None,
+    ) -> None:
+        self._client = client or AsyncClient(
+            api_key=api_key,
+            base_url=base_url,
+            http_client=httpx.AsyncClient(
+                timeout=5.0,
+                follow_redirects=True,
+                limits=httpx.Limits(
+                    max_connections=1000,
+                    max_keepalive_connections=100,
+                    keepalive_expiry=120,
+                ),
+            ),
+        )
+        self._assistant_opts = assistant_opts
+        self._running_fncs: MutableSet[asyncio.Task[Any]] = set()
+
+        self._sync_openai_task = asyncio.create_task(self._sync_openai())
+
+    async def _sync_openai(self) -> AssistantLoadOptions:
+        if isinstance(self._assistant_opts, AssistantCreateOptions):
+            assistant = await self._client.beta.assistants.create(
+                model=self._assistant_opts.model,
+                name=self._assistant_opts.name,
+                instructions=self._assistant_opts.instructions,
+                tools=[{"type": t} for t in self._assistant_opts.tools],
+            )
+            thread = await self._client.beta.threads.create()
+            return AssistantLoadOptions(assistant_id=assistant.id, thread_id=thread.id)
+        else:
+            return self._assistant_opts
+
+    def chat(
+        self,
+        *,
+        chat_ctx: llm.ChatContext,
+        fnc_ctx: llm.FunctionContext | None = None,
+        temperature: float | None = None,
+        n: int | None = 1,
+        parallel_tool_calls: bool | None = None,
+    ):
+        return AssistantLLMStream(
```

Comment:
would there be race issues if a user does:

```
llm.chat(message1)
llm.chat(message2)
```

is it possible for message1 to override the message context from message2 there? should the `chat` function limit only a single active stream at any given time?

Reply:
This is a limitation of the OpenAI api. For any given thread_id, only one stream can be running at a time. Also, messages can not be mutated while that stream is running. The llm instance represents the thread. 

Logically it makes more sense for the chat_ctx to be the object that represents a thread but since anyone can mutate and/or copy the chat_ctx at any time (VoiceAssistant), we needed additional logic and state which couldn't really live anywhere else.